src/irmin-pack/pack_key_intf.ml

```diff
@@ -17,19 +17,60 @@
 open! Import
 
 module type S = sig
-  type hash
-
-  include Irmin.Key.S with type t = hash and type hash := hash
+  include Irmin.Key.S
 
   val null : t
+
+  val unfindable_of_hash : hash -> t
+  (** [unfindable_of_hash h] is a key [k] such that [to_hash k = h], with an
+      unspecified internal representation. This function enables an efficient
+      implmentation of "portable" inodes, but is otherwise unused. Attempting to
+      dereference a key constructed in this way results in undefined behaviour. *)
```

Comment:
Rather than undefined behaviour, can we throw an exception when dereferencing?

Reply:
It is a bit mysterious to me how one can go from a hash to a key.

Reply:
As mentioned here (https://github.com/mirage/irmin/pull/1659#discussion_r765835475), this is done only for performance when computing a hash of an inode whose children have not yet been added to the store (to avoid needing first to allocate a separate in-memory object with `s/key/hash` in the appropriate places). The Inode interface ensures that this trick is kept only to portable inodes (which can't be saved). If we _did_ ever try to dereference one, an exception will still be raised (since these keys reference values with "length" of -1), but it'll be a messy one.